src/util/RandomEvictionCache.h

```diff
@@ -0,0 +1,201 @@
+        mValuePtrs.pop_back();
+        ++mCounters.mEvicts;
+    }
+
+  public:
+    explicit RandomEvictionCache(size_t maxSize) : mMaxSize(maxSize)
+    {
+        mValueMap.reserve(maxSize + 1);
+        mValuePtrs.reserve(maxSize + 1);
+    }
+
+    size_t
+    maxSize() const
+    {
+        return mMaxSize;
+    }
+
+    size_t
+    size() const
+    {
+        return mValueMap.size();
+    }
+
+    Counters const&
+    getCounters() const
+    {
+        return mCounters;
+    }
+
+    void
+    put(K const& k, V const& v)
+    {
+        ++mGeneration;
+        CacheValue newValue{mGeneration, v};
+        auto pair = mValueMap.insert(std::make_pair(k, newValue));
+        if (pair.second)
+        {
+            // An insertion occurred: save a pointer to the pair inserted.
+            // Note that it's okay to do so, as the C++14 standard
+            // guarantees that rehashing does not invalidate pointers or
+            // references to elements (26.2.7)
+            map_value_type& inserted = *pair.first;
+            mValuePtrs.push_back(&inserted);
+            ++mCounters.mInserts;
+            // We may have just grown over the size limit. Fix that.
+            if (mValuePtrs.size() > mMaxSize)
+            {
+                evictOne();
+            }
+        }
+        else
+        {
+            // No insertion happened, was already an entry: update its value.
+            CacheValue& existing = pair.first->second;
+            existing = newValue;
+            ++mCounters.mUpdates;
+        }
+    }
+
+    bool
+    exists(K const& k, bool countMisses = true)
+    {
+        bool miss = (mValueMap.find(k) == mValueMap.end());
+        // We do not count hits here; but we usually count misses, as exists()
+        // is typically used as a guard followed by a get(). If you're using it
+        // _not_ that way, pass `false` for countMisses.
+        //
+        // Or use the maybeGet interface, which will save you a second hash
+        // lookup and provides a less-cumbersome interface.
+        if (miss && countMisses)
+        {
+            ++mCounters.mMisses;
+        }
+        return !miss;
+    }
+
+    void
+    clear()
+    {
+        mValuePtrs.clear();
+        mValueMap.clear();
+    }
+
+    template <typename F>
+    void
+    erase_if(F const& f)
+    {
+        for (size_t i = 0; i < mValuePtrs.size(); ++i)
+        {
+            map_value_type*& vp = mValuePtrs.at(i);
+            while (mValuePtrs.size() != 0 && f(vp->second.mValue))
+            {
+                mValueMap.erase(vp->first);
+                std::swap(vp, mValuePtrs.back());
+                mValuePtrs.pop_back();
+            }
+        }
+    }
+
+    V&
+    get(K const& k)
+    {
+        if (auto* p = maybeGet(k))
+        {
+            return *p;
+        }
+        else
+        {
+            throw std::range_error("There is no such key in cache");
+        }
+    }
+
+    V*
+    maybeGet(K const& k)
```

Comment:
I'm not sure about this interface to the cache, I would probably feel more comfortable with something like `bool maybeGet(K const& k, V& v)`. The reason for this is that we often use caches of smart pointers where exists-in-cache-as-nullptr means it doesn't exist in the underlying data store (this is in `LedgerTxn` for example). This means that we'll have to write and review code like
```
auto v = maybeGet(k); // This should really say vptr here, but it's easy to forget
if (v && *v) // Easy to forget the second condition
{
// Do something that depends on *v being non-null 
}
```
where-as the other interface would be easier to use correctly with
```
V v;
if (maybeGet(k, v) && v) // Clear what each condition does, and that you really need both
{
// Do something that depends on v being non-null 
}
```

I'm not extremely determined about this change, but I do think it would be an improvement. What do @marta-lokhova and @graydon think?

Reply:
It does seem like it'd be easy to make mistakes with the current state of things. In addition to what you said, we probably shouldn't give a raw pointer to the client in general, as we don't know when they will dereference it, and if the data it points to is still valid at that time.  

I'm wondering if we should remove `maybeGet` entirely. While `exists` + `get` is a little less efficient (two map lookups instead of one), it is also less error-prone. And hits/misses will still be updated properly, as `exists` will count misses if configured correctly.

Reply:
Agreed that the `maybeGet` with a bool and out-param is probably a nicer interface. Don't especially care if you want to eliminate it altogether, I just have a long history of people yelling at me when designing container APIs that make you do two lookups, but I agree it's probably not a big deal.